### PR TITLE
Migrate to doctrine/lexer v3.0

### PR DESCRIPTION
**Pull request by Ph0tonic**
[Doctrine/L](https://github.com/doctrine/lexer) just released a brand new version 3.0 and here is a PR to use this newer version.

Let me know @egulias if I need to change anything, I am not familiar with psalm annotations.

### CODE REVIEW

File: src/EmailLexer.php
Comment (egulias) [inline src/EmailLexer.php]: You can remove this annotation, given `$previous` is now a proper type

File: src/EmailLexer.php
Comment (egulias) [inline src/EmailLexer.php]: You can remove this annotation, given `$previous` is now a proper type

File: src/EmailLexer.php
Comment (egulias) [inline src/EmailLexer.php]: Should this also be `Token`?

File: src/EmailLexer.php
Comment (egulias) [inline src/EmailLexer.php]: You can remove this annotation, given `$previous` is now a proper type.

File: src/EmailLexer.php
Comment (egulias) [inline src/EmailLexer.php]: Replace with `Token`

File: src/EmailLexer.php
Comment (egulias) [inline src/EmailLexer.php]: Type is `Token` here, right?

File: src/MessageIDParser.php
Comment (egulias) [inline src/MessageIDParser.php]: `$this->lexer->token->value`

**Review by egulias — CHANGES_REQUESTED** [review]
Hey, thanks for taking the time. Very nice PR to update the dependency, at last it brings a `Token` type.
I have made a bunch of comments but please take a close look to Scrutinizer comments on the files. There are methods in the new version of the Lexer library that have types different to the one in the code.
Also check the _Checks_ tab and see why tests are failing.

Looking forward to the changes. And thanks again for the PR!

**Comment by Ph0tonic** [comment]
Hey @egulias ,

Thank for your feedback, I applied them and fixed the tests. There was a type check issue due to the value `null` of the lex `S_EMPTY`. I changed it to `-1` and made the relevant changes to make it work.

I also finally managed to fix all Psalm issues so I'll let you review the changes 👍

**Comment by Pasqualle** [comment]
doctrine/lexer v3.0.0 is also available now:
https://packagist.org/packages/doctrine/lexer

**Comment by Ph0tonic** [comment]
> doctrine/lexer v3.0.0 is also available now: https://packagist.org/packages/doctrine/lexer

Thank for the info @Pasqualle 

@egulias The main difference with this new v3.0.0 version is the dropped support for PHP <8.1. From what I have read in your readme, we should create a new v4.x release branch.
I updated PHP versions in Travis, with the newer supported version but I wasn't sure about the support of ppc64le architecture. Let me know what you think about that. 👍

**Comment by derrabus** [comment]
Note that Lexer 3 is Lexer 2 minus the deprecation layer, so requiring Lexer 3 without allowing v2 as well is not really necessary. Lexer 3 is not supported by Doctrine ORM 2 and it likely won't ever be. This basically means that bumping to `"doctrine/lexer": "^3"` instead of `"doctrine/lexer": "^2 || ^3"` would forbid the use of this package in a project that also uses Doctrine ORM.

**Comment by derrabus** [comment]
I've submitted #340 to unblock Lexer 2 (and doctrine/annotations 2) in Symfony's CI.

**Comment by egulias** [comment]
Thanks @derrabus, very useful.
Will merge #340 and then upgrade this library to v4 with this code.
Will work on this during the week.

@Ph0tonic I might need you to do a PR to another branch for this. Will let you know.

Thanks guys!

**Comment by Ph0tonic** [comment]
> Thanks @derrabus, very useful. Will merge #340 and then upgrade this library to v4 with this code. Will work on this during the week.
> 
> @Ph0tonic I might need you to do a PR to another branch for this. Will let you know.
> 
> Thanks guys!

Great thank you, let me know if you need me to do anything.

File: README.md
Comment (egulias) [inline README.md]: Can you please mark v2 as EOL -> YES
And keep release date for 4 clear for now, work in progress!

File: composer.json
Comment (egulias) [inline composer.json]: Please keep the original formatting :)

File: composer.json
Comment (egulias) [inline composer.json]: Please upgrade to latest, 1.26

File: .travis.yml
Comment (egulias) [inline .travis.yml]: Update to Jammy please, this was set like this to be able to build PHP 7.

**Review by egulias — CHANGES_REQUESTED** [review]
We are almost there! A couple of more changes please.
I'll create the 4.x branch and will request you to do the PR to that branch, once we have sorted out all the changes on this PR (so keep the conversation in one place).

Thanks!

**Comment by Ph0tonic** [comment]
Great thank you for your feedback. Here are the requested changes. 👍

File: composer.json
Comment (derrabus) [inline composer.json]: Again, I would strongly recommend to loosen this constraint to…

```suggestion
    "doctrine/lexer": "^2.0 || ^3.0",
```

This should be possible without much extra effort and avoid dependency conflicts downstream for the time being.

Comment (egulias) [reply composer.json]: I have checked Doctrine's ORM and v3 is requiring v2 of the lexer, so @Ph0tonic can you please update to 
```suggestion
    "doctrine/lexer": "^2 || ^3",
```
Thanks @derrabus

Comment (Ph0tonic) [reply composer.json]: Done 👍

**Comment by egulias** [comment]
@Ph0tonic please close this PR and create a new one against v4 branch: https://github.com/egulias/EmailValidator/tree/4.x .
I'll take care of the conflicts.
Thnkas!

**Comment by egulias** [comment]
Closing in favor of #344

**Comment by derrabus** [comment]
You mean #344, I guess. 🙃

**Comment by egulias** [comment]
Updated, thanks!

**Pull request closed without merging**
